### PR TITLE
Run E2E tests with Travis

### LINKED ISSUE #721: CI: Automate e2e environment setup in docker containers
Currently we can run e2e tests only against preconfigured environment.
Automate environment setup to run e2e tests in CI task.
Available options to consider:

* [@woocommerce/e2e-environment](https://github.com/woocommerce/woocommerce/tree/update/e2e-hosting/tests/e2e/env)
* [@wordpress/env](https://github.com/WordPress/gutenberg/tree/master/packages/env)
* Own script reusing existing client and server docker setups.

### PR DESCRIPTION
**Pull request by vbelolapotkov**
Fixes #721 

#### Changes proposed in this Pull Request

* Setup environment for e2e tests with a script.
* Run e2e tests with `@wordpress-scripts`.
* Update travis config to run e2e tests on each PR.

#### Testing instructions

**Locally** - follow instructions in `tests/e2e/README.md`

**In Travis** - check output of "E2E tests" build step. It should pass successfully.

Successful run: https://github.com/Automattic/woocommerce-payments/pull/746/commits/27f21bb740e5932acc5dbc7ebcd6e581319ea3a1 | [travis details](https://github.com/Automattic/woocommerce-payments/runs/850481415)
Failed run: https://github.com/Automattic/woocommerce-payments/pull/746/commits/8d3e7c83e2613af17a7beb4d392949b3a8166ebd | [travis details](https://github.com/Automattic/woocommerce-payments/runs/850655268)
Slack report: p1594226655417300-slack-CQ0Q6N62D

**TODO:**

- [x] Update Readme.
- [x] Fix failing test in travis. 
- [x] Restore unit tests in travis and run only on master. 
-------------------

- [x] Tested on mobile (or does not apply)

### CODE REVIEW

File: tests/e2e/README.md [hunk 1]
```diff
@@ -1,17 +1,35 @@
...
-At the moment E2E tests can be run only locally.
+E2E tests can be run locally or in Travis CI.
 
 ## Setup
 
-1. Make sure the site is up and running locally, including Jetpack connection setup, test products and initial WC Payments plugin configuration.
-2. Update config file `tests/e2e/config/default.json` or create `test.json` file in the same folder with required params. For local testing create file `local.json` (ignored by version control) with `url` param matching local site under test, e.g. `http://localhost:8082`.
+Setup script requires the following env variables to be configured:
+```
+WCP_SERVER_REPO='http://github.com/server-repo.git'
```
[the remark this answers is not in the document]
Comment (v18) [reply tests/e2e/README.md]: To follow up here, the error I get is:

```
remote: Repository not found.
fatal: repository 'https://github.com/Automattic/server-repo.git/' not found
```

It works with a url of the form `git@github.com:org/server-repo.git`. Not sure what the difference might be between our set ups.

File: tests/e2e/README.md [hunk 2]
```diff
@@ -1,17 +1,35 @@
...
+E2E_WCPAY_STRIPE_TEST_CLIENT_ID=<stripe ca_xxx>
+E2E_WCPAY_STRIPE_TEST_WEBHOOK_SIGNATURE_KEY=<stripe whsec_xxx>
+E2E_WCPAY_STRIPE_ACCOUNT_ID=<stripe acct_id>
+
+// Optional to see verbose output
+DEBUG=true
+```
+
+Create file `local.env` in the repository root folder to setup variables locally.
+For local environment make sure to run `npm install`,  `composer install` and `npm run build:client` before running setup script.
```
[the remark this answers is not in the document]
Comment (v18) [reply tests/e2e/README.md]: Understood 👍

**Comment by v18** [comment]
I took another look @vbelolapotkov. Great work, I like how this is turning out.

I think this is close to ready to merge. Here are some notes and questions for your consideration:

1 - (Not really a blocker) I was able to run `npm run test:e2e` successfully. When running `npm run test:e2e-dev` it finished successfully visually (I saw the "Order received" text), but there was a timeout error:

```
Timeout - Async callback was not invoked within the 5000ms timeout specified by jest.setTimeout
```

2 - (Not a blocker) When running both `npm run test:e2e` and `npm run test:e2e-dev` I get this warning:

```
  console.error node_modules/config/lib/config.js:1430
    WARNING: NODE_ENV value of 'test' did not match any deployment config file names.

  console.error node_modules/config/lib/config.js:1431
    WARNING: See https://github.com/lorenwest/node-config/wiki/Strict-Mode
```

I don't see any NODE_ENV defined with a value of 'test' in our files, so this might be coming from the e2e library (I haven't checked).

3 - I noticed we switched from using the woocommerce e2e testing library to `@wordpress/e2e-test-utils`. What's spurred this change?

I think it would be great to link to the docs you used for how to set up the e2e tests using this library in the README, so that others can follow along if they want or need to.

4 - Do we have internal documentation for where to look out for Slack notifications? And for viewing screenshots from Travis?

5 - (Not a blocker) I'd be curious to hear about what the problem was with Travis in the beginning and how it was resolved 😊 

👏 👏

File: tests/e2e/README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (vbelolapotkov) [reply tests/e2e/README.md]: Usually this happens when git is missing auth info for https protocol. Anyway, added both options with https://github.com/Automattic/woocommerce-payments/pull/746/commits/24de034d4c6f7d1d9062c716175d6901fde86307

**Comment by vbelolapotkov** [comment]
> I took another look @vbelolapotkov. Great work, I like how this is turning out.

Thanks for taking another round and feedback! 🙂
 
> 1 - (Not really a blocker) I was able to run `npm run test:e2e` successfully. When running `npm run test:e2e-dev` it finished successfully visually (I saw the "Order received" text), but there was a timeout error:
> 
> ```
> Timeout - Async callback was not invoked within the 5000ms timeout specified by jest.setTimeout
> ```

Hm, not sure where it comes from. And 5000ms value looks unfamiliar to me.

> 2 - (Not a blocker) When running both `npm run test:e2e` and `npm run test:e2e-dev` I get this warning:
> 
> ```
>   console.error node_modules/config/lib/config.js:1430
>     WARNING: NODE_ENV value of 'test' did not match any deployment config file names.
> 
>   console.error node_modules/config/lib/config.js:1431
>     WARNING: See https://github.com/lorenwest/node-config/wiki/Strict-Mode
> ```
> 
> I don't see any NODE_ENV defined with a value of 'test' in our files, so this might be coming from the e2e library (I haven't checked).

It's set [here](https://github.com/WordPress/gutenberg/blob/818d0458c3fa8845694e57247ed06d135a581a31/packages/scripts/scripts/test-e2e.js#L3) by `@wordpress/scripts` and the warnings are generated by `config` package.

I've renamed config file name in https://github.com/Automattic/woocommerce-payments/pull/746/commits/f817c6d547fc45e622f2f424b486c54c9d4fa226 to match `NODE_ENV`.

> 3 - I noticed we switched from using the woocommerce e2e testing library to `@wordpress/e2e-test-utils`. What's spurred this change?
> 

That's a long story here and I'm going to write a p2 post about it but here are the key points: 

* There are several woo e2e related packages and we used only configs from woo/e2e-environment which didn't give us many benefits and rather created more problems (e.g. with stripe iframes not available in dev mode). I used configs from @automattic/puppeteer-utils instead and slightly adopted them. This package was advertised as base configs set for woo e2e as well but somehow wasn't used by `woo/e2e-environment`.
* I haven't switched `@wordpress/e2e-test-utils`. This one was introduced before and used by utility functions, e.g. [here](https://github.com/Automattic/woocommerce-payments/pull/746/files#diff-80082fcde5e84942223068e9ee048bcaR8).

> I think it would be great to link to the docs you used for how to set up the e2e tests using this library in the README, so that others can follow along if they want or need to.

I haven't used any particular docs except package README's and src code, so not sure where to link from here. But I'm going to share e2e setup experience in that p2 post. Stay tuned 🙂

> 
> 4 - Do we have internal documentation for where to look out for Slack notifications? And for viewing screenshots from Travis?

Great point! IMO FG page is a good place for such an update. Added to my TODO list.


> 5 - (Not a blocker) I'd be curious to hear about what the problem was with Travis in the beginning and how it was resolved 😊

That's a great question 🙂 The answer is `host.docker.internal` hostname is not available in docker in linux environment used by travis and the fix is here https://github.com/Automattic/woocommerce-payments/pull/746/commits/97cddeb757023a0cf1631c627a6881b98b9fd6a8 .

**Comment by v18** [comment]
@vbelolapotkov thanks for explanations! From my perspective, this is ready to merge. Going to mark as ready now. But also wanted to ping @jrodger (👋 ) -  in case there are some more things you'd like to chime in on.

**Comment by jrodger** [comment]
Doh, my test runs started failing again. Dev mode:
```
 FAIL  tests/e2e/specs/purchase.spec.js
  ● Test suite failed to run

    connect ECONNREFUSED 127.0.0.1:59227

```

Headless:
```
 FAIL  tests/e2e/specs/purchase.spec.js (46.569s)
  Successful purchase
    ✕ successful purchase (30700ms)

  ● Successful purchase › successful purchase

    TimeoutError: Navigation Timeout Exceeded: 20000ms exceeded

      138 |             await Promise.all( [
      139 |                     expect( page ).toClick( '#place_order' ),
    > 140 |                     page.waitForNavigation( { waitUntil: 'networkidle0' } ),
          |                          ^
      141 |             ] );
      142 |     },
      143 | 

```

I'd be happy with merging this in and seeing if anyone else has the same issue, rather than holding things up because there's something up with my local setup.

What's the plan for merging? This PR is set to merge into `add/e2e-tests` currently. Do we need a bit of a tidy up before merging to `master`?

**Comment by vbelolapotkov** [comment]
> connect ECONNREFUSED 127.0.0.1:59227

Oh, that's something weird. I'd try to stop containers nuke `tests/e2e/deps` and `tests/e2e/docker`. Run fresh setup `npm run test:e2e-setup` and then `npm run test:e2e` or `npm run test:e2e-dev` (both headless and dev should work now.

> What's the plan for merging? This PR is set to merge into add/e2e-tests currently. Do we need a bit of a tidy up before merging to master?

Let's merge this first. Give it a couple of runs on `add/e2e-tests` to make sure it works 🙂 And then make a final review/merge.

**Comment by jrodger** [comment]
> Let's merge this first. Give it a couple of runs on add/e2e-tests to make sure it works 🙂 And then make a final review/merge.

Sounds like a plan :shipit:

**Pull request merged**